### PR TITLE
Copter:　Issue/#6357

### PR DESCRIPTION
**Pull request by kouseii**
This fixes this issue #6357.

### CODE REVIEW

**Comment by kouseii** [comment]
Should I make duplicate code a function?

**Comment by kouseii** [comment]
"_cd" may be the meaning of centi degrees.
I'll change the param names "yaw".

**Comment by rmackay9** [comment]
This is really great.  A few comments:

- there is some duplicate state we should think about combining: guided_angle_state.yaw_rate_cds and guided_yaw_state.yaw_rate
- auto_yaw_mode vs guided_yaw_state.use_yaw-rate

Some style things:

- commits should be of the form "Copter: xxx"
- before the final merge we should "squash" the commits together.  If you do this it's ok to do a "forced" push to this branch.

Again, this is a really good contribution.  I hope we can pull it into master in the next week or so.

**Comment by kouseii** [comment]
Dear Randy.

Thank you for your comment.
I combined guided_yaw_state into guided_angle_state.
But I feel complicated about auto_yaw_mode.
In actuality,
if both yaw and yaw_rate were specified, I want to invalidate yaw_rate when it reaches yaw.
And implementation of Copter :: set_auto_yaw_look_at_heading () is still in progress.

File: ArduCopter/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -1750,11 +1750,11 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
         }
 
         if (!pos_ignore && !vel_ignore && acc_ignore) {
```
Comment (muramura) [inline ArduCopter/GCS_Mavlink.cpp]: I think that you can judge acc_ignore alone.

File: ArduCopter/control_guided.cpp [hunk 2]
```diff
@@ -23,8 +23,9 @@ struct {
     float yaw_cd;
     float yaw_rate_cds;
     float climb_rate_cms;
+    bool use_yaw;
     bool use_yaw_rate;
-} static guided_angle_state = {0,0.0f, 0.0f, 0.0f, 0.0f, 0.0f, false};
+} static guided_angle_state = {0,0.0f, 0.0f, 0.0f, 0.0f, 0.0f, false, false};
```
Comment (muramura) [inline ArduCopter/control_guided.cpp]: White space: I think you should put it in.

File: ArduCopter/control_guided.cpp [hunk 3]
```diff
@@ -427,8 +440,13 @@ void Copter::guided_pos_control_run()
         // roll & pitch from waypoint controller, yaw rate from pilot
         attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(wp_nav->get_roll(), wp_nav->get_pitch(), target_yaw_rate, get_smoothing_gain());
     }else{
```
Comment (muramura) [inline ArduCopter/control_guided.cpp]: White space: I think you should put it in.

Comment (kouseii) [inline ArduCopter/control_guided.cpp]: I'll fix it.

File: ArduCopter/control_guided.cpp (hunk 2 above)
Comment (kouseii) [inline ArduCopter/control_guided.cpp]: I'll fix it.

File: ArduCopter/GCS_Mavlink.cpp (hunk 1 above)
Comment (kouseii) [inline ArduCopter/GCS_Mavlink.cpp]: I don't like adding new functionality and refactoring at the same time.
Remembers Kent Beck's words. "Only Wear One Of Four Hats."

File: ArduCopter/control_guided.cpp [hunk 4]
```diff
@@ -554,9 +577,14 @@ void Copter::guided_posvel_control_run()
...
         attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(pos_control->get_roll(), pos_control->get_pitch(), target_yaw_rate, get_smoothing_gain());
-    }else{
-        // roll, pitch from waypoint controller, yaw heading from auto_heading()
-        attitude_control->input_euler_angle_roll_pitch_yaw(pos_control->get_roll(), pos_control->get_pitch(), get_auto_heading(), true, get_smoothing_gain());
+    } else {
+        if (guided_angle_state.use_yaw_rate) {
+            // roll & pitch from waypoint controller, yaw rate from GCS
+            attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(pos_control->get_roll(), pos_control->get_pitch(), guided_angle_state.yaw_rate_cds, get_smoothing_gain());
+        } else {
+            // roll, pitch from waypoint controller, yaw heading from auto_heading()
```
Comment (muramura) [inline ArduCopter/control_guided.cpp]: I think that comments are not covered by processing.

File: ArduCopter/GCS_Mavlink.cpp (hunk 1 above)
Comment (muramura) [inline ArduCopter/GCS_Mavlink.cpp]: I understand.

Comment (kouseii) [inline ArduCopter/GCS_Mavlink.cpp]: Thank you for your understanding.

File: ArduCopter/control_guided.cpp (hunk 4 above)
Comment (kouseii) [inline ArduCopter/control_guided.cpp]: my fault!
I'll fix it.

**Comment by rmackay9** [comment]
Hi there.  I haven't completed tested but I have squashed your commits and then added my own changes on top in this branch: https://github.com/rmackay9/rmackay9-ardupilot/commits/kouseii-6357

I'm not sure if it's better or worse but I think what I've done is more consistent with the structure for handling the yaw control.  There should be no functional change to your original implementation.

Instead of extending the guided_angle_state (which is currently used in response to a SET_ATTITUDE_TARGET message), I've extended the auto_yaw_mode with a new AUTO_YAW_RATE.

**Comment by rmackay9** [comment]
I'm closing this in favour of a new PR that incorporates the changes from this PR.

**Pull request closed without merging**